File: 09_recursion/05_palindrome_partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        self.res = []
        self.s = s
        self.n = len(s)
        self.helper()
        return self.res

    def helper(self, ind=0, curr=[]):
        if ind == self.n:
            self.res.append(curr[:])
            return

        for i in range(ind, self.n):
            curr_string = self.s[ind : i + 1]
            if curr_string == curr_string[::-1]:
                curr.append(curr_string)
                self.helper(i + 1, curr)
                curr.pop()

File: 09_recursion/05_palindrome_partitioning.rs
impl Solution {
    pub fn partition(s: String) -> Vec<Vec<String>> {
        let mut res: Vec<Vec<String>> = Vec::new();
        let mut curr: Vec<String> = Vec::new();
        Solution::helper(0, &mut curr, &s, &mut res);
        res
    }

    fn helper(ind: usize, curr: &mut Vec<String>, s: &String, res: &mut Vec<Vec<String>>) {
        if ind == s.len() {
            res.push(curr.clone());
            return;
        }

        for i in ind + 1..=s.len() {
            if s[ind..i] == s[ind..i].chars().rev().collect::<String>() {
                curr.push(s[ind..i].to_string());
                Solution::helper(i, curr, s, res);
                curr.pop();
            }
        }
    }
}
